src/v/cluster/topic_table.cc

```diff
@@ -40,26 +40,31 @@ topic_table::transform_topics(Func&& f) const {
 topic_table::topic_metadata::topic_metadata(
   topic_configuration_assignment c, model::revision_id rid) noexcept
   : configuration(std::move(c))
-  , _id_or_topic(rid) {}
+  , _source_topic(std::nullopt)
+  , _revision(rid) {}
 
 topic_table::topic_metadata::topic_metadata(
-  topic_configuration_assignment c, model::topic st) noexcept
+  topic_configuration_assignment c,
+  model::revision_id rid,
+  model::topic st) noexcept
   : configuration(std::move(c))
-  , _id_or_topic(std::move(st)) {}
+  , _source_topic(st)
+  , _revision(rid) {}
 
 bool topic_table::topic_metadata::is_topic_replicable() const {
-    return std::holds_alternative<model::revision_id>(_id_or_topic);
+    return _source_topic.has_value() == false;
 }
+
 model::revision_id topic_table::topic_metadata::get_revision() const {
-    vassert(
-      is_topic_replicable(), "Query for revision_id on a non-replicable topic");
-    return std::get<model::revision_id>(_id_or_topic);
+    return _revision;
```

Comment:
@graphcareful with the assertion removed what safety checks do we have?

Reply:
I restored this check. It won't hurt to have it.